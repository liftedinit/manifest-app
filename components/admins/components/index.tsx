export * from './validatorList';
export * from './stakeHolderPayout';
export * from './chainUpgrader';
