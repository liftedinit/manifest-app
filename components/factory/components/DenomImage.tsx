import { MetadataSDKType } from '@liftedinit/manifestjs/dist/codegen/cosmos/bank/v1beta1/bank';
import Image from 'next/image';
import { useEffect, useState } from 'react';

import { ProfileAvatar } from '@/utils/identicon';

export const supportedDomains = [
  'imgur.com',
  'i.imgur.com',
  'cloudfront.net',
  'cdn.jsdelivr.net',
  'raw.githubusercontent.com',
  's3.amazonaws.com',
  'storage.googleapis.com',
  'res.cloudinary.com',
  'images.unsplash.com',
  'media.giphy.com',
  'media.istockphoto.com',
  'imgix.net',
  'staticflickr.com',
  'twimg.com',
  'pinimg.com',
  'giphy.com',
  'dropboxusercontent.com',
  'googleusercontent.com',
  'upload.wikimedia.org',
  'unsplash.com',
  'istockphoto.com',
  't4.ftcdn.net',
];

export const supportedPatterns = [
  /^https:\/\/.*\.s3\.amazonaws\.com/,
  /^https:\/\/.*\.storage\.googleapis\.com/,
  /^https:\/\/.*\.cloudinary\.com/,
  /^https:\/\/.*\.imgix\.net/,
  /^https:\/\/.*\.staticflickr\.com/,
  /^https:\/\/.*\.twimg\.com/,
  /^https:\/\/.*\.pinimg\.com/,
  /^https:\/\/.*\.giphy\.com/,
  /^https:\/\/.*\.dropboxusercontent\.com/,
  /^https:\/\/.*\.googleusercontent\.com/,
  /^https:\/\/.*\.unsplash\.com/,
  /^https:\/\/.*\.upload\.wikimedia\.org/,
  /^https:\/\/.*\.istockphoto\.com/,
  /^https:\/\/.*\.media\.giphy\.com/,
  /^https:\/\/.*\.media\.istockphoto\.com/,
  /^https:\/\/.*\.images\.unsplash\.com/,
  /^https:\/\/.*\.media\.istockphoto\.com/,
  /^https:\/\/.*\.imgix\.net/,
  /^https:\/\/.*\.staticflickr\.com/,
  /^https:\/\/.*\.twimg\.com/,
  /^https:\/\/.*\.pinimg\.com/,
  /^https:\/\/.*\.giphy\.com/,
  /^https:\/\/.*\.t4\.ftcdn\.net/,
];

export const isUrlSupported = (url: string) => {
  try {
    const { hostname } = new URL(url);
    return (
      supportedDomains.includes(hostname) || supportedPatterns.some(pattern => pattern.test(url))
    );
  } catch {
    return false;
  }
};

export const DenomImage = ({
  denom,
  withBackground = true,
}: {
  denom?: MetadataSDKType | null;
  withBackground?: boolean;
}) => {
  const [imageError, setImageError] = useState(false);
  const [isSupported, setIsSupported] = useState(false);

  useEffect(() => {
    if (denom?.uri) {
      setIsSupported(isUrlSupported(denom.uri));
    }
  }, [denom?.uri]);

  // Check for MFX token first
  if (denom?.base === 'umfx' || denom?.base?.includes('uosmo')) {
    return (
      <div
        className={`w-11 h-11 p-2 rounded-md ${withBackground ? 'dark:bg-[#ffffff0f] bg-[#0000000a]' : ''}`}
      >
        <Image
          width={44}
          height={44}
          src={denom?.base === 'umfx' ? '/logo.svg' : '/osmosis.svg'}
          alt="MFX Token Icon"
          className="w-full h-full data-[loaded=false]:animate-pulse data-[loaded=false]:skeleton data-[loaded=false]:bg-gray-300"
        />
      </div>
    );
  }

  // Then check for other conditions
  if (!denom?.uri || !isSupported || imageError) {
    return <ProfileAvatar walletAddress={denom?.base} withBackground={withBackground} />;
  }

  // For all other cases, use the denom.uri
  return (
    <div
      className={`w-11 h-11 p-2 rounded-md ${withBackground ? 'dark:bg-[#ffffff0f] bg-[#0000000A]' : ''}`}
    >
      <Image
        width={44}
        height={44}
        src={denom?.uri}
        alt="Token Icon"
        onError={() => setImageError(true)}
        className="rounded-md w-full h-full data-[loaded=false]:animate-pulse data-[loaded=false]:skeleton data-[loaded=false]:bg-gray-300"
      />
    </div>
  );
};
