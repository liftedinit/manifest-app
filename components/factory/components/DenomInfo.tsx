import { TruncatedAddressWithCopy } from '@/components/react/addressCopy';
import { shiftDigits } from '@/utils';
import { MetadataSDKType } from '@liftedinit/manifestjs/dist/codegen/cosmos/bank/v1beta1/bank';
import { DenomInfoModal } from '../modals';
import { UpdateDenomMetadataModal } from '../modals';
import { CoinSDKType } from '@liftedinit/manifestjs/dist/codegen/cosmos/base/v1beta1/coin';

export default function DenomInfo({
  denom,
  address,
  refetchDenoms,
  balance,
  isBalanceLoading,
}: Readonly<{
  denom: MetadataSDKType | null;
  address: string;
  refetchDenoms: () => void;
  balance: CoinSDKType | null;
  isBalanceLoading: boolean;
}>) {
  const DenomConversion = ({ denom }: { denom: MetadataSDKType }) => {
    if (!denom || !denom.denom_units || denom.denom_units.length === 0) {
      return (
        <div className="text-md">
          1&nbsp;{0}&nbsp;=&nbsp;{1}
          &nbsp;{1}
        </div>
      );
    }

    const displayUnit = denom.display;
    const baseUnit = denom.denom_units[0].denom.split('/').pop();
    const conversionFactor = 10 ** denom.denom_units[1]?.exponent;

    return (
      <div className="text-md max-w-[20ch] truncate">
        1&nbsp;{displayUnit}&nbsp;=&nbsp;{conversionFactor.toLocaleString()}
        &nbsp;{baseUnit?.toUpperCase()}
      </div>
    );
  };

  const baseUnit = denom?.denom_units[0].denom.split('/').pop();

  return (
    <div className="flex flex-col max-h-[23rem] relative shadow min-h-[23rem] rounded-md bg-base-100 w-full p-4">
      <div className="w-full rounded-md">
        <div className="px-4 py-2 justify-between items-center border-base-content">
          <div className="flex flex-row w-full justify-between -mt-[0.1rem] items-center">
            <h3 className="text-lg font-bold leading-6">Metadata</h3>
            <button
              onClick={() => {
                const modal = document.getElementById(
                  `update_metadata_${denom?.base}_${denom?.name}`
                ) as HTMLDialogElement;
                modal?.showModal();
              }}
              className="btn btn-primary btn-xs"
              disabled={denom?.base.includes('mfx')}
              aria-label="update metadata"
            >
              Update
            </button>
          </div>
          <div className="divider divider-horizon -mt-0"></div>
        </div>
        {!denom && (
          <div className="p-4 py-24 -mt-4 underline text-center">
            <p>No Denom Selected</p>
          </div>
        )}
        {denom && (
          <div className="flex flex-col">
            <div className="flex flex-col gap-3 justify-left px-4 -mt-4 rounded-md items-left">
              <span className="text-sm leading-3 capitalize text-gray-400">TICKER</span>
              <span className="text-xl  max-w-[30ch] truncate" aria-label="ticker">
                {denom?.display ?? 'No Ticker available'}
              </span>
            </div>

            <div className="flex gap-4 px-4 flex-row py-4 rounded-md">
              <div className="flex flex-col w-1/2">
                <div className="flex flex-col gap-2 bg-base-300 p-4 rounded-md justify-left mb-6 items-left">
                  <span className="text-sm capitalize text-gray-400 truncate">BASE DENOM</span>
                  <div aria-label="base denom">
                    <TruncatedAddressWithCopy address={denom.base} slice={14} />
                  </div>
                </div>

                <div className="flex flex-col gap-2 bg-base-300 p-4 rounded-md justify-left items-left">
                  <span className="text-sm capitalize text-gray-400 md:block hidden">
                    YOUR BALANCE
                  </span>
                  <span className="text-sm capitalize text-gray-400 block md:hidden">BALANCE</span>
                  <div className="flex flex-row gap-1 animate-fadeIn items-center justify-start truncate">
                    {isBalanceLoading ? (
                      <span className="text-md animate-fadeIn">
                        0 &nbsp;
                        {baseUnit?.toUpperCase().slice(1)}
                      </span>
                    ) : (
                      <span className="text-md animate-fadeIn">
                        {shiftDigits(balance?.amount ?? '0', -denom?.denom_units[1]?.exponent ?? 6)}
                        &nbsp;
                        {baseUnit?.toUpperCase().slice(1)}
                      </span>
                    )}
                  </div>
                </div>
              </div>
              <div className="flex flex-col w-1/2">
                <div className="flex flex-col bg-base-300 p-4 rounded-md gap-2 justify-left mb-6 items-left">
                  <span className="text-sm capitalize text-gray-400 truncate">EXPONENT</span>

                  <DenomConversion denom={denom} />
                </div>
                <div className="flex flex-col bg-base-300 p-4 rounded-md gap-2 justify-left items-left">
                  <span className="text-sm capitalize text-gray-400">SYMBOL</span>
                  <div
                    className="flex flex-row justify-between items-start max-w-[30ch] truncate"
                    aria-label="symbol"
                  >
                    {denom?.symbol ?? 'No Description'}
                    <div className="flex-row justify-between items-center gap-2 hidden md:flex">
                      <button
                        onClick={() => {
                          const modal = document.getElementById(
                            `denom_info_${denom.base}`
                          ) as HTMLDialogElement;
                          modal?.showModal();
                        }}
                        className="btn btn-secondary btn-xs"
                      >
                        More Info
                      </button>
                    </div>
                  </div>

                  {denom && (
                    <>
                      <UpdateDenomMetadataModal
                        key={`update_modal_${denom?.base}_${denom?.name}`}
                        denom={denom}
                        address={address}
                        modalId={`update_metadata_${denom?.base}_${denom?.name}`}
                        onSuccess={refetchDenoms}
                      />
                      <DenomInfoModal denom={denom} modalId={`denom_info_${denom?.base}`} />
                    </>
                  )}
                </div>
              </div>
            </div>
          </div>
        )}
      </div>
    </div>
  );
}
