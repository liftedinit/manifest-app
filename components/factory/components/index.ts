export * from './DenomList';
export * from './DenomImage';
