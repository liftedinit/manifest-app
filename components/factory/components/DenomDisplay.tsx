import { DenomImage, VerifiedIcon } from '@/components';
import { formatTokenDisplay } from '@/utils';
import React from 'react';

export const DenomDisplay = ({
  denom,
  metadata,
  withBackground,
}: {
  metadata?: {
    display?: string;
    base?: string;
  } | null;
  denom?: string;
  withBackground?: boolean;
}) => {
  const name = formatTokenDisplay(denom ?? metadata?.display ?? '?').toUpperCase();
  const verified = metadata?.base === 'umfx';

  return (
    <>
      <div className="flex items-center justify-center">
        <DenomImage withBackground={withBackground} denom={metadata} />
      </div>
      <p className="align-middle font-semibold text-[#161616] dark:text-white">
        {name}
        {verified && <VerifiedIcon className="w-4 mx-1 inline relative bottom-1 text-primary" />}
      </p>
    </>
  );
};
