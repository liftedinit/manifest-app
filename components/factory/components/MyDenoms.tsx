import { useState, useEffect, useMemo } from 'react';
import { useRouter } from 'next/router';
import { DenomImage } from './DenomImage';
import Link from 'next/link';
import { truncateString } from '@/utils';
import { SearchIcon, MintIcon, BurnIcon } from '@/components/icons';
import { DenomInfoModal } from '@/components/factory/modals/denomInfo';
import MintModal from '@/components/factory/modals/MintModal';
import BurnModal from '@/components/factory/modals/BurnModal';
import { UpdateDenomMetadataModal } from '@/components/factory/modals/updateDenomMetadata';
import { PiInfo } from 'react-icons/pi';
import { ExtendedMetadataSDKType, shiftDigits } from '@/utils';
import { MultiMintModal } from '@/components/factory/modals';
import { MultiBurnModal } from '../modals/multiMfxBurnModal';
import { usePoaGetAdmin } from '@/hooks';
import useIsMobile from '@/hooks/useIsMobile';

export default function MyDenoms({
  denoms,
  isLoading,
  refetchDenoms,
  address,
}: {
  denoms: ExtendedMetadataSDKType[];
  isLoading: boolean;
  refetchDenoms: () => void;
  address: string;
}) {
  const [searchQuery, setSearchQuery] = useState('');
  const [currentPage, setCurrentPage] = useState(1);
  const [openUpdateDenomMetadataModal, setOpenUpdateDenomMetadataModal] = useState(false);
  const isMobile = useIsMobile();

  const pageSize = isMobile ? 6 : 8;

  const router = useRouter();
  const [selectedDenom, setSelectedDenom] = useState<ExtendedMetadataSDKType | null>(null);
  const [modalType, setModalType] = useState<
    'mint' | 'burn' | 'multimint' | 'multiburn' | 'update' | 'info' | null
  >(null);

  const filteredDenoms = useMemo(() => {
    return denoms.filter(denom => denom?.display.toLowerCase().includes(searchQuery.toLowerCase()));
  }, [denoms, searchQuery]);

  const totalPages = Math.ceil(filteredDenoms.length / pageSize);
  const paginatedDenoms = filteredDenoms.slice(
    (currentPage - 1) * pageSize,
    currentPage * pageSize
  );

  const handleDenomSelect = (denom: ExtendedMetadataSDKType) => {
    if (!modalType) {
      setSelectedDenom(denom);
      setModalType('info');
      router.push(`/factory?denom=${denom.base}&action=info`, undefined, { shallow: true });
    }
  };

  const { poaAdmin, isPoaAdminLoading } = usePoaGetAdmin();

  useEffect(() => {
    const { denom, action } = router.query;
    if (denom && typeof denom === 'string') {
      const decodedDenom = decodeURIComponent(denom);
      const metadata = denoms.find(d => d.base === decodedDenom);
      if (metadata) {
        setSelectedDenom(metadata);
        if (
          action === 'mint' ||
          action === 'burn' ||
          action === 'multimint' ||
          action === 'multiburn' ||
          action === 'update' ||
          action === 'info'
        ) {
          setModalType(action as 'mint' | 'burn' | 'multimint' | 'multiburn' | 'update' | 'info');
          if (action === 'update') {
            setOpenUpdateDenomMetadataModal(true);
          }
        } else {
          setModalType('info');
        }
      }
    } else {
      setSelectedDenom(null);
      setModalType(null);
    }
  }, [router.query, denoms]);

  const handleCloseModal = () => {
    setSelectedDenom(null);
    setModalType(null);
    setOpenUpdateDenomMetadataModal(false);
    router.push('/factory', undefined, { shallow: true });
  };

  const handleUpdateModalClose = () => {
    setSelectedDenom(null);
    setOpenUpdateDenomMetadataModal(false);
    setModalType(null);
    router.push('/factory', undefined, { shallow: true });
  };

  const handleUpdateModal = (denom: ExtendedMetadataSDKType, e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    setSelectedDenom(denom);
    setModalType('update');
    setOpenUpdateDenomMetadataModal(true);
    router.push(`/factory?denom=${denom.base}&action=update`, undefined, { shallow: true });
  };

  const handleSwitchToMultiMint = () => {
    setModalType('multimint');
    router.push(`/factory?denom=${selectedDenom?.base}&action=multimint`, undefined, {
      shallow: true,
    });
  };

  const handleSwitchToMultiBurn = () => {
    setModalType('multiburn');
    router.push(`/factory?denom=${selectedDenom?.base}&action=multiburn`, undefined, {
      shallow: true,
    });
  };

  return (
    <div className="relative w-full h-screen overflow-hidden">
      <div className="h-full flex flex-col p-4">
        <div className="flex flex-col md:flex-row justify-between items-start md:items-center mb-4 gap-4">
          <div className="flex flex-col sm:flex-row items-start sm:items-center gap-4 w-full md:w-auto">
            <h1
              className="text-secondary-content"
              style={{ fontSize: '20px', fontWeight: 700, lineHeight: '24px' }}
            >
              My Factory
            </h1>
            <div className="relative w-full sm:w-[224px]">
              <input
                type="text"
                placeholder="Search for a token..."
                className="input input-bordered w-full h-[40px] rounded-[12px] border-none bg-secondary text-secondary-content pl-10 focus:outline-none focus-visible:ring-1 focus-visible:ring-primary"
                value={searchQuery}
                onChange={e => setSearchQuery(e.target.value)}
              />
              <SearchIcon className="h-6 w-6 absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400" />
            </div>
          </div>

          <div className="hidden md:block">
            <Link href="/factory/create" passHref>
              <button className="btn btn-gradient w-[224px] h-[52px] text-white rounded-[12px] focus:outline-none focus-visible:ring-1 focus-visible:ring-primary">
                Create New Token
              </button>
            </Link>
          </div>
        </div>
        <div className="overflow-auto">
          <div className="max-w-8xl mx-auto">
            <table className="table w-full border-separate border-spacing-y-3">
              <thead className="sticky top-0 bg-[#F0F0FF] dark:bg-[#0E0A1F]">
                <tr className="text-sm font-medium">
                  <th className="bg-transparent w-1/4">Token</th>
                  <th className="bg-transparent w-2/5 xl:table-cell hidden">Name</th>
                  <th className="bg-transparent w-2/5  md:table-cell hidden">Total Supply</th>
                  <th className="bg-transparent w-1/4">Actions</th>
                </tr>
              </thead>
              <tbody className="space-y-4">
                {isLoading
                  ? Array(isMobile ? 6 : 8)
                      .fill(0)
                      .map((_, index) => (
                        <tr key={index}>
                          <td className="bg-secondary rounded-l-[12px] w-1/4">
                            <div className="flex items-center space-x-3">
                              <div
                                className="skeleton w-10 h-8 rounded-full shrink-0"
                                aria-label={`skeleton-${index}-avatar`}
                              />
                              <div
                                className="skeleton font-medium xxs:max-xs:hidden block"
                                aria-label={`skeleton-${index}-name`}
                              />
                            </div>
                          </td>
                          <td className="bg-secondary w-2/5 xl:table-cell hidden">
                            <div
                              className="skeleton h-2 w-8"
                              aria-label={`skeleton-${index}-symbol`}
                            />
                          </td>
                          <td className="bg-secondary w-2/5 sm:w-1/4">
                            <div
                              className="skeleton h-2 w-24"
                              aria-label={`skeleton-${index}-supply`}
                            />
                          </td>
                          <td className="bg-secondary rounded-r-[12px] w-1/4">
                            <div className="flex space-x-2">
                              <button
                                className="btn btn-md btn-outline btn-square btn-primary"
                                disabled
                              >
                                <MintIcon className="w-7 h-7 text-current opacity-50" />
                              </button>
                              <button
                                className="btn btn-md btn-outline btn-square btn-error"
                                disabled
                              >
                                <BurnIcon className="w-7 h-7 text-current opacity-50" />
                              </button>
                              <button
                                className="btn btn-md btn-outline btn-square btn-info"
                                disabled
                              >
                                <PiInfo className="w-7 h-7 text-current opacity-50" />
                              </button>
                            </div>
                          </td>
                        </tr>
                      ))
                  : paginatedDenoms.map(denom => (
                      <TokenRow
                        key={denom.base}
                        denom={denom}
                        onSelectDenom={() => handleDenomSelect(denom)}
                        onMint={e => {
                          e.stopPropagation();
                          setSelectedDenom(denom);
                          setModalType('mint');
                          router.push(`/factory?denom=${denom.base}&action=mint`, undefined, {
                            shallow: true,
                          });
                        }}
                        onBurn={e => {
                          e.stopPropagation();
                          setSelectedDenom(denom);
                          setModalType('burn');
                          router.push(`/factory?denom=${denom.base}&action=burn`, undefined, {
                            shallow: true,
                          });
                        }}
                        onUpdate={e => handleUpdateModal(denom, e)}
                      />
                    ))}
              </tbody>
            </table>
            {totalPages > 1 && (
              <div
                className="flex items-center justify-center gap-2 "
                onClick={e => e.stopPropagation()}
                role="navigation"
                aria-label="Pagination"
              >
                <button
                  onClick={e => {
                    e.stopPropagation();
                    setCurrentPage(prev => Math.max(1, prev - 1));
                  }}
                  disabled={currentPage === 1 || isLoading}
                  className="p-2 hover:bg-[#0000001A] dark:hover:bg-[#FFFFFF1A] text-black dark:text-white rounded-lg transition-colors disabled:opacity-50 disabled:cursor-not-allowed"
                  aria-label="Previous page"
                >
                  ‹
                </button>

                {[...Array(totalPages)].map((_, index) => {
                  const pageNum = index + 1;
                  if (
                    pageNum === 1 ||
                    pageNum === totalPages ||
                    (pageNum >= currentPage - 1 && pageNum <= currentPage + 1)
                  ) {
                    return (
                      <button
                        key={pageNum}
                        onClick={e => {
                          e.stopPropagation();
                          setCurrentPage(pageNum);
                        }}
                        className={`w-8 h-8 flex items-center justify-center rounded-lg transition-colors text-black dark:text-white 
                          ${currentPage === pageNum ? 'bg-[#0000001A] dark:bg-[#FFFFFF1A]' : 'hover:bg-[#0000001A] dark:hover:bg-[#FFFFFF1A]'}`}
                        aria-label={`Page ${pageNum}`}
                        aria-current={currentPage === pageNum ? 'page' : undefined}
                      >
                        {pageNum}
                      </button>
                    );
                  } else if (pageNum === currentPage - 2 || pageNum === currentPage + 2) {
                    return (
                      <span key={pageNum} aria-hidden="true">
                        ...
                      </span>
                    );
                  }
                  return null;
                })}

                <button
                  onClick={e => {
                    e.stopPropagation();
                    setCurrentPage(prev => Math.min(totalPages, prev + 1));
                  }}
                  disabled={currentPage === totalPages || isLoading}
                  className="p-2 hover:bg-[#0000001A] dark:hover:bg-[#FFFFFF1A] text-black dark:text-white rounded-lg transition-colors disabled:opacity-50 disabled:cursor-not-allowed"
                  aria-label="Next page"
                >
                  ›
                </button>
              </div>
            )}
          </div>
          <div className="block md:hidden mt-8">
            <Link href="/factory/create" passHref>
              <button className="btn btn-gradient w-full h-[52px] text-white rounded-[12px]">
                Create New Token
              </button>
            </Link>
          </div>
        </div>
      </div>
      <DenomInfoModal
        openDenomInfoModal={modalType === 'info'}
        setOpenDenomInfoModal={open => {
          if (!open) {
            handleCloseModal();
          }
        }}
        denom={selectedDenom}
        modalId="denom-info-modal"
      />
      <MintModal
        admin={poaAdmin ?? 'manifest1afk9zr2hn2jsac63h4hm60vl9z3e5u69gndzf7c99cqge3vzwjzsfmy9qj'}
        isPoaAdminLoading={isPoaAdminLoading}
        denom={modalType === 'mint' ? selectedDenom : null}
        address={address}
        refetch={refetchDenoms}
        balance={selectedDenom?.balance ?? '0'}
        totalSupply={selectedDenom?.totalSupply ?? '0'}
        isOpen={modalType === 'mint'}
        onClose={handleCloseModal}
        onSwitchToMultiMint={handleSwitchToMultiMint}
      />
      <BurnModal
        denom={selectedDenom}
        address={address}
        refetch={refetchDenoms}
        balance={selectedDenom?.balance ?? '0'}
        totalSupply={selectedDenom?.totalSupply ?? '0'}
        isOpen={modalType === 'burn'}
        onClose={handleCloseModal}
        onSwitchToMultiBurn={handleSwitchToMultiBurn}
      />
      <UpdateDenomMetadataModal
        modalId="update-denom-metadata-modal"
        denom={selectedDenom}
        address={address}
        onSuccess={() => {
          refetchDenoms();
          handleUpdateModalClose();
        }}
        openUpdateDenomMetadataModal={openUpdateDenomMetadataModal}
        setOpenUpdateDenomMetadataModal={open => {
          if (!open) {
            handleUpdateModalClose();
          } else {
            setOpenUpdateDenomMetadataModal(true);
          }
        }}
      />
      <MultiMintModal
        admin={poaAdmin ?? 'manifest1afk9zr2hn2jsac63h4hm60vl9z3e5u69gndzf7c99cqge3vzwjzsfmy9qj'}
        address={address}
        denom={selectedDenom}
        exponent={selectedDenom?.denom_units[1]?.exponent ?? 0}
        refetch={refetchDenoms}
        isOpen={modalType === 'multimint'}
        onClose={handleCloseModal}
      />
      <MultiBurnModal
        admin={poaAdmin ?? 'manifest1afk9zr2hn2jsac63h4hm60vl9z3e5u69gndzf7c99cqge3vzwjzsfmy9qj'}
        address={address}
        denom={selectedDenom}
        exponent={selectedDenom?.denom_units[1]?.exponent ?? 0}
        refetch={refetchDenoms}
        isOpen={modalType === 'multiburn'}
        onClose={handleCloseModal}
      />
    </div>
  );
}

function TokenRow({
  denom,
  onSelectDenom,
  onMint,
  onBurn,
  onUpdate,
}: {
  denom: ExtendedMetadataSDKType;
  onSelectDenom: () => void;
  onMint: (e: React.MouseEvent) => void;
  onBurn: (e: React.MouseEvent) => void;
  onUpdate: (e: React.MouseEvent) => void;
}) {
  // Add safety checks for the values
  const exponent = denom?.denom_units?.[1]?.exponent ?? 0;
  const totalSupply = denom?.totalSupply ?? '0';
  const balance = denom?.balance ?? '0';

  // Format numbers safely
  const formatAmount = (amount: string) => {
    try {
      return Number(shiftDigits(amount, -exponent)).toLocaleString(undefined, {
        maximumFractionDigits: exponent,
      });
    } catch (error) {
      console.warn('Error formatting amount:', error);
      return '0';
    }
  };

  return (
    <tr
      className="group text-black dark:text-white rounded-lg cursor-pointer transition-colors"
      onClick={onSelectDenom}
    >
      <td className="bg-secondary group-hover:bg-base-300 rounded-l-[12px] w-1/4">
        <div className="flex items-center space-x-3">
          <DenomImage denom={denom} />
          <span className="font-medium xxs:max-xs:hidden block">
            {denom.display.startsWith('factory')
              ? denom.display.split('/').pop()?.toUpperCase()
              : truncateString(denom.display, 12)}
          </span>
        </div>
      </td>
      <td className="bg-secondary group-hover:bg-base-300 w-2/5 xl:table-cell hidden">
        {truncateString(denom?.name ?? 'No name provided', 20)}
      </td>
      <td className="bg-secondary group-hover:bg-base-300 w-2/5 md:table-cell hidden sm:w-1/4">
        <div className="flex flex-col sm:flex-row sm:items-center ">
          <span className="sm:mr-2">{formatAmount(totalSupply)}</span>
          <span className="font-extralight">
            {truncateString(denom?.display ?? 'No ticker provided', 10).toUpperCase()}
          </span>
        </div>
      </td>
      <td
        className="bg-secondary group-hover:bg-base-300 rounded-r-[12px] w-1/4"
        onClick={e => e.stopPropagation()}
      >
        <div className="flex space-x-2">
          <button
            className="btn btn-md bg-base-300 text-primary btn-square group-hover:bg-secondary hover:outline hover:outline-primary hover:outline-1 outline-none"
            onClick={onMint}
          >
            <MintIcon className="w-7 h-7 text-current" />
          </button>

          <button
            disabled={denom.base.includes('umfx')}
            className="btn btn-md bg-base-300 text-primary btn-square group-hover:bg-secondary hover:outline hover:outline-primary hover:outline-1 outline-none"
            onClick={onBurn}
          >
            <BurnIcon className="w-7 h-7 text-current" />
          </button>

          <button
            disabled={denom.base.includes('umfx')}
            className="btn btn-md bg-base-300 text-primary btn-square group-hover:bg-secondary hover:outline hover:outline-primary hover:outline-1 outline-none"
            onClick={e => {
              e.preventDefault();
              e.stopPropagation();
              onUpdate(e);
            }}
          >
            <PiInfo className="w-7 h-7 text-current" />
          </button>
        </div>
      </td>
    </tr>
  );
}
