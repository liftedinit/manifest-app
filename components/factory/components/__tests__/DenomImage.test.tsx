import { MetadataSDKType } from '@liftedinit/manifestjs/dist/codegen/cosmos/bank/v1beta1/bank';
import { cleanup, render, screen, waitFor } from '@testing-library/react';
import { Mock, afterEach, beforeEach, describe, expect, mock, spyOn, test } from 'bun:test';

import { DenomImage } from '@/components/factory/components/DenomImage';
import { clearAllMocks, formatComponent, mockModule } from '@/tests';

// A cute little candle gif
const uri =
  'https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExdHg1aHVqa3NoMG45bzYwbHR5ODk3b2JqbHhnemlmcXpjOXB0enExMSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/zHcirZSkw8RSDhawFl/giphy.gif';

const renderWithProps = (props = {}) => {
  const defaultProps = {
    denom: {
      base: 'umfx',
      uri: uri,
    } as MetadataSDKType,
  };
  return render(<DenomImage {...defaultProps} {...props} />);
};

describe('DenomImage', () => {
  beforeEach(() => {
    mockModule('next/image', () => ({
      default: (props: any) => {
        // eslint-disable-next-line @next/next/no-img-element,jsx-a11y/alt-text
        return <img alt="" {...props} />;
      },
    }));
  });

  afterEach(() => {
    clearAllMocks();
    cleanup();
  });

  test('renders MFX token image correctly', async () => {
    const mockup = renderWithProps();
    await waitFor(() => expect(document.querySelector('[src=/logo.svg]')).toBeInTheDocument());
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot();
  });

  test('renders ProfileAvatar for unsupported URL', async () => {
    const mockup = renderWithProps({
      denom: { base: 'unsupported', uri: 'https://unsupported.com/token.png' },
    });
    await waitFor(() => expect(screen.getByAltText('Profile Avatar')).toBeInTheDocument());
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot();
  });

  test('renders image from supported URL', async () => {
    const mockup = renderWithProps({ denom: { base: 'supported', uri: uri } });
    await waitFor(() => expect(document.querySelector(`[src="${uri}"]`)).toBeInTheDocument());
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot();
  });
});
