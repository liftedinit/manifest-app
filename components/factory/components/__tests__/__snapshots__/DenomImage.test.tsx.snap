// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`DenomImage renders MFX token image correctly 1`] = `
"<DocumentFragment>
  <div
    class="w-11 h-11 p-2 rounded-md dark:bg-[#ffffff0f] bg-[#0000000a]"
  >
    <img
      alt="MFX Token Icon"
      class="w-full h-full data-[loaded=false]:animate-pulse data-[loaded=false]:skeleton data-[loaded=false]:bg-gray-300"
      height="44"
      src="/logo.svg"
      width="44"
    />
  </div>
</DocumentFragment>"
`;

exports[`DenomImage renders ProfileAvatar for unsupported URL 1`] = `
"<DocumentFragment>
  <img
    alt="Profile Avatar"
    class="rounded-md dark:bg-[#FFFFFF0F] bg-[#0000000A]"
    src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHdpZHRoPSc3MDAnIGhlaWdodD0nNzAwJyBzdHlsZT0nYmFja2dyb3VuZC1jb2xvcjpyZ2JhKDI1NSwyNTUsMjU1LDApOyc+PGcgc3R5bGU9J2ZpbGw6cmdiYSg4OCwyMTgsMjEwLDEpOyBzdHJva2U6cmdiYSg4OCwyMTgsMjEwLDEpOyBzdHJva2Utd2lkdGg6My41Oyc+PHJlY3QgIHg9JzMwOScgeT0nMTQ3JyB3aWR0aD0nODEnIGhlaWdodD0nODEnLz48cmVjdCAgeD0nMzA5JyB5PScyMjgnIHdpZHRoPSc4MScgaGVpZ2h0PSc4MScvPjxyZWN0ICB4PSczMDknIHk9JzM5MCcgd2lkdGg9JzgxJyBoZWlnaHQ9JzgxJy8+PHJlY3QgIHg9JzIyOCcgeT0nMjI4JyB3aWR0aD0nODEnIGhlaWdodD0nODEnLz48cmVjdCAgeD0nMzkwJyB5PScyMjgnIHdpZHRoPSc4MScgaGVpZ2h0PSc4MScvPjxyZWN0ICB4PScxNDcnIHk9JzMwOScgd2lkdGg9JzgxJyBoZWlnaHQ9JzgxJy8+PHJlY3QgIHg9JzQ3MScgeT0nMzA5JyB3aWR0aD0nODEnIGhlaWdodD0nODEnLz48cmVjdCAgeD0nMTQ3JyB5PSczOTAnIHdpZHRoPSc4MScgaGVpZ2h0PSc4MScvPjxyZWN0ICB4PSc0NzEnIHk9JzM5MCcgd2lkdGg9JzgxJyBoZWlnaHQ9JzgxJy8+PC9nPjwvc3ZnPg=="
    style="height: 44px; width: 44px;"
  />
</DocumentFragment>"
`;

exports[`DenomImage renders image from supported URL 1`] = `
"<DocumentFragment>
  <div
    class="w-11 h-11 p-2 rounded-md dark:bg-[#ffffff0f] bg-[#0000000A]"
  >
    <img
      alt="Token Icon"
      class="rounded-md w-full h-full data-[loaded=false]:animate-pulse data-[loaded=false]:skeleton data-[loaded=false]:bg-gray-300"
      height="44"
      src="https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExdHg1aHVqa3NoMG45bzYwbHR5ODk3b2JqbHhnemlmcXpjOXB0enExMSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/zHcirZSkw8RSDhawFl/giphy.gif"
      width="44"
    />
  </div>
</DocumentFragment>"
`;
